File: client/src/lib/sync.ts
/**
 * One sync: push what is waiting, then catch up. Called on app open, on
 * regaining connectivity, and after every movement (FR-OFF-03).
 *
 * Push comes first so a deactivated account's last work lands before the
 * server refuses it (FR-OFF-06). Every response re-measures the clock offset
 * (NFR-DATA-13) and its round trip; the round trip is sent back with the next
 * push so the server can allow for latency in its own measurement.
 */
import { type Api, ApiError, Offline } from "./api";
import { uploadPhotos } from "./photos";
import { Store } from "./store";

export type SyncOutcome = { ok: true } | { ok: false; reason: "offline" | "signed_out" | "error"; message: string };

export async function sync(store: Store, api: Api, now: () => number = Date.now): Promise<SyncOutcome> {
  if (!store.meta.token) return { ok: false, reason: "signed_out", message: "not signed in" };
  try {
    const pending = store.pending;
    let unidentified = 0;
    // A log_id on this push's reply that differs from ours means the server's database was
    // replaced since we last synced: our cursor is meaningless there, whatever it says.
    let replaced = false;
    const stampRoundTrip = (offset: number | undefined, round_trip: number) =>
      store.setMeta({ round_trip_ms: round_trip, ...(offset !== undefined ? { clock_offset: offset } : {}) });
    if (pending.length > 0) {
      const { data, offset, round_trip } = await api.push(
        store.meta.device_id,
        now(),
        pending.map((e) => Store.outgoing(e)),
        store.meta.round_trip_ms,
      );
      await stampRoundTrip(offset, round_trip);
      if (store.meta.log_id !== undefined && data.log_id !== store.meta.log_id) replaced = true;
      const pushed = await store.pushed(data.accepted, data.rejected);
      unidentified += pushed.unidentified;
      // A sequence collision (two tabs, or old data from before it was fixed) is not a real
      // rejection: the affected events were re-stamped and are ready to go up again, once.
      if (pushed.retry && store.pending.length > 0) {
        const {
          data: retried,
          offset: retriedOffset,
          round_trip: retriedRoundTrip,
        } = await api.push(
          store.meta.device_id,
          now(),
          store.pending.map((e) => Store.outgoing(e)),
          store.meta.round_trip_ms,
        );
        await stampRoundTrip(retriedOffset, retriedRoundTrip);
        if (!replaced && store.meta.log_id !== undefined && retried.log_id !== store.meta.log_id) replaced = true;
        const retriedPushed = await store.pushed(retried.accepted, retried.rejected);
        unidentified += retriedPushed.unidentified;
      }
    }
    // Photos taken offline go up now, before pull brings back the events that name them (FR-INV-11).
    await uploadPhotos(store, api);

    // A cursor with no log_id was stored by an older build of this app. It may
    // belong to a database that has since been replaced, so it is not trusted.
    if (replaced || store.meta.cursor === undefined || store.meta.log_id === undefined) {
      await bootstrap(store, api);
    } else {
      try {
        await pull(store, api);
      } catch (error) {
        // Our cursor means nothing to this server any more: too old, or the
        // database was restored. Start again from a snapshot (FR-OFF-14).
        if (error instanceof ApiError && error.status === 410) await bootstrap(store, api);
        else throw error;
      }
    }

    await store.setMeta({ last_sync_at: now() });
    await store.trim(now());
    // The rest of sync still ran — the device is not stuck — but the server could not match some
    // rejections to anything it holds, so report the batch as unclean rather than as a full success.
    if (unidentified > 0) {
      return {
        ok: false,
        reason: "error",
        message: `the server refused ${unidentified} record(s) it could not identify`,
      };
    }
    return { ok: true };
  } catch (error) {
    if (error instanceof Offline) return { ok: false, reason: "offline", message: error.message };
    if (error instanceof ApiError) {
      if (error.round_trip !== undefined || error.offset !== undefined) {
        await store.setMeta({
          ...(error.round_trip !== undefined ? { round_trip_ms: error.round_trip } : {}),
          ...(error.offset !== undefined ? { clock_offset: error.offset } : {}),
        });
      }
      if (error.status === 401 || error.code === "deactivated") {
        await store.setMeta({ token: undefined, user: undefined });
        return { ok: false, reason: "signed_out", message: error.message };
      }
      return { ok: false, reason: "error", message: error.message };
    }
    throw error;
  }
}

async function bootstrap(store: Store, api: Api): Promise<void> {
  const { data, offset, round_trip } = await api.bootstrap();
  // The snapshot, cursor and log_id all belong together (see Store.bootstrap); the offset and
  // round trip are only a convenience and are set separately, and only when the server gave us
  // an offset to use.
  await store.bootstrap(data.snapshot, data.cursor, data.log_id);
  await store.setMeta({
    round_trip_ms: round_trip,
    ...(offset !== undefined ? { clock_offset: offset } : {}),
    // The current upcoming-events list, so it works offline too (FR-RES-20). Not part of the
    // snapshot: the server owns it and replaces it whole, there is no history to replay.
    calendar_events: data.calendar_events,
  });
}

/** The device calls again until it gets an empty page. */
async function pull(store: Store, api: Api): Promise<void> {
  for (;;) {
    const since = store.meta.cursor ?? 0;
    const { data, offset, round_trip } = await api.pull(since, store.meta.log_id);
    await store.setMeta({
      round_trip_ms: round_trip,
      ...(offset !== undefined ? { clock_offset: offset } : {}),
      calendar_events: data.calendar_events,
    });
    // A non-empty page must move the cursor past where we asked from; one that does not is a
    // server bug, not something a retry fixes, so this stops the loop instead of spinning forever.
    if (data.events.length > 0 && data.cursor <= since) {
      throw new Error(`pull returned events but the cursor did not advance past ${since} (got ${data.cursor})`);
    }
    await store.receive(data.events, data.cursor);
    if (data.events.length === 0) return;
  }
}

File: src/gear_tracker/sync.py
"""The three sync operations, with no HTTP in them. app.py puts HTTP in front.

See docs/architecture.md, "Sync". Every result carries server_time so the
device can re-measure its clock offset (NFR-DATA-13).
"""

from __future__ import annotations

import logging
import sqlite3
from dataclasses import asdict, dataclass
from typing import Annotated, Any

from pydantic import Field, ValidationError

from gear_tracker import derived, events
from gear_tracker.errors import ApiError, BadRequest, Deactivated, Forbidden, Rebootstrap
from gear_tracker.events import NonEmpty, Rejected, Strict, now_ms
from gear_tracker.flags import add_flag

logger = logging.getLogger(__name__)

SyncError = ApiError

RETENTION_MS = 90 * 24 * 3_600_000
"""How far back a device keeps history (NFR-DATA-03). A cursor older than this re-bootstraps instead."""

DRIFT_THRESHOLD_MS = 60_000
"""A stored clock offset this far from a fresh measurement means the device clock moved. Flag, do not trust."""

PAGE_SIZE = 1000


@dataclass(frozen=True)
class Principal:
    """Who is calling. M4 builds one from a credential; until then tests build them directly."""

    user_id: str
    device_id: str
    active: bool = True
    role: str = "user"


class PushBody(Strict):
    device_id: NonEmpty
    client_time: int
    events: list[Any]
    """Each event is judged on its own, so a bad one is a rejection, not a 400."""
    round_trip_ms: Annotated[int, Field(ge=0)] | None = None
    """The device's last measured round trip (client/src/lib/clock.ts measureOffset halves this to
    get its own offset). Omitted by an older client, or before the device has ever synced."""


def _require_active(principal: Principal) -> None:
    if not principal.active:
        raise Deactivated("this account has been deactivated")


def log_id(conn: sqlite3.Connection) -> str:
    """This log's identity, set when the database was created (migration 0007).

    A device holds the one its snapshot came from. A different one means a
    different log, whatever the cursor says.
    """
    return str(conn.execute("SELECT value FROM meta WHERE key = 'log_id'").fetchone()[0])


def bootstrap(conn: sqlite3.Connection, principal: Principal, now: int | None = None) -> dict[str, Any]:
    """Current state and the cursor it was true at, read in one transaction (FR-OFF-14)."""
    _require_active(principal)
    now = now_ms() if now is None else now
    conn.execute("BEGIN")
    try:
        snapshot = derived.snapshot(conn)
        cursor = derived.cursor(conn)
    finally:
        conn.execute("COMMIT")
    return {"snapshot": snapshot, "cursor": cursor, "log_id": log_id(conn), "server_time": now}


def push(conn: sqlite3.Connection, principal: Principal, body: Any, now: int | None = None) -> dict[str, Any]:
    """Take a batch of events from one device. Idempotent on event id.

    A deactivated account is still allowed here, and only here (FR-OFF-06).
    Every event is attributed to the signed-in user and device; the server
    does not take the device's word for either.
    """
    now = now_ms() if now is None else now
    try:
        batch = PushBody.model_validate(body)
    except ValidationError as exc:
        raise BadRequest(Rejected.from_validation(exc).reason) from None
    if batch.device_id != principal.device_id:
        raise Forbidden("device_id does not match the credential")

    # The client's own offset is a round trip halved, so its recorded clock_offset already allows
    # for half the latency of the sync that measured it. A one-way `now - client_time` here has no
    # such allowance and overstates the offset by the full latency of this push; halving the round
    # trip this push took (as last measured) puts the two on the same footing.
    half_trip = batch.round_trip_ms // 2 if batch.round_trip_ms is not None else 0
    measured_offset = now - batch.client_time - half_trip
    accepted: list[str] = []
    rejected: list[dict[str, Any]] = []
    for incoming in batch.events:
        if not isinstance(incoming, dict):
            rejected.append({"id": None, "reason": "event must be a JSON object"})
            continue
        try:
            if incoming.get("actor_id") != principal.user_id:
                raise Rejected("actor_id must be the signed-in user")
            if incoming.get("device_id") != principal.device_id:
                raise Rejected("device_id must be this device")
            _check_entity_rules(conn, principal, incoming)
            seen_before = events.get(conn, incoming["id"]) is not None if isinstance(incoming.get("id"), str) else False
            stored = events.append(conn, incoming, received_at=now)
            if not seen_before:
                _check_drift(conn, stored, measured_offset, now)
            accepted.append(stored.id)
        except Rejected as exc:
            rejected.append({"id": incoming.get("id"), "reason": exc.reason})
            logger.warning(
                "rejected event %s (%s on %s/%s) from device %s: %s",
                incoming.get("id"),
                incoming.get("type"),
                incoming.get("entity_type"),
                incoming.get("entity_id"),
                principal.device_id,
                exc.reason,
            )
        except Exception:
            # A bad one is a rejection, not a 500 for the whole batch. events.append rolls back
            # its own work on the way out, so the connection is clean for the next event.
            rejected.append({"id": incoming.get("id"), "reason": "the server could not store this event"})
            logger.exception(
                "event %s (%s on %s/%s) from device %s could not be stored",
                incoming.get("id"),
                incoming.get("type"),
                incoming.get("entity_type"),
                incoming.get("entity_id"),
                principal.device_id,
            )
    return {"accepted": accepted, "rejected": rejected, "log_id": log_id(conn), "server_time": now}


def _is_pool(conn: sqlite3.Connection, item_id: Any) -> bool:
    """Whether an id, if it names anything, names a pool (FR-INV-34). A missing or non-item id is not one."""
    if not isinstance(item_id, str):
        return False
    item = derived.get_entity(conn, "item", item_id)
    return bool(item and item.get("pool"))


def _check_entity_rules(conn: sqlite3.Connection, principal: Principal, incoming: dict[str, Any]) -> None:
    """What a device may not do, whatever it says. Field checks happen later, in events.validate."""
    entity_type, kind = incoming.get("entity_type"), incoming.get("type")
    if entity_type == "user":
        raise Rejected("user changes go through the accounts API")
    if entity_type == "setting" and principal.role != "admin":
        raise Rejected("settings are changed by an Admin")
    if entity_type == "location" and principal.role != "admin":
        raise Rejected("locations are an Admin's job (FR-SET-05)")
    if entity_type == "category" and principal.role != "admin" and kind != "created":
        # Anyone signed in may add a category, same as the item editor (see assistant.add_category).
        raise Rejected("categories are renamed and deleted by an Admin (FR-SET-05)")
    if entity_type == "item" and kind == "field_changed":
        payload = incoming.get("payload")
        field = payload.get("field") if isinstance(payload, dict) else None
        # Deleting takes a record off every list for good (FR-INV-32), so it stays with an Admin.
        if field == "deleted" and principal.role != "admin":
            raise Rejected("items are deleted by an Admin")
        # A pool has no units (FR-INV-34): moving one under a pool is refused, whether at creation or later.
        if field == "parent_id" and _is_pool(conn, payload.get("value")):
            raise Rejected("a pool has no units (FR-INV-34)")
        if field == "merged_into" and principal.role != "admin":
            raise Rejected("items are merged by an Admin (FR-INV-13)")
    if entity_type == "item" and kind == "created":
        payload = incoming.get("payload")
        if isinstance(payload, dict):
            if payload.get("pool") and not payload.get("generic"):
                raise Rejected("a pool must be generic (FR-INV-34)")
            if _is_pool(conn, payload.get("parent_id")):
                raise Rejected("a pool has no units (FR-INV-34)")
    if entity_type == "item" and kind in ("checked_out", "checked_in"):
        item = derived.get_entity(conn, "item", str(incoming.get("entity_id"))) or {}
        if item.get("deleted"):
            raise Rejected("this item was deleted")
        is_pool = bool(item.get("pool"))
        if item.get("generic") and not is_pool:
            raise Rejected("a generic item does not move; its units do (FR-INV-21)")
        payload = incoming.get("payload")
        has_count = isinstance(payload, dict) and payload.get("count") is not None
        if is_pool and not has_count:
            raise Rejected("a pool moves by count")
        if not is_pool and has_count:
            raise Rejected("count is only for a pool (FR-OUT-22)")
        if kind == "checked_out":
            if item.get("retired"):
                raise Rejected("retired items cannot be checked out (FR-INV-04)")
            if item.get("merged_into"):
                raise Rejected("this item was merged into another (FR-INV-13)")
    if entity_type == "item" and kind == "recounted":
        item = derived.get_entity(conn, "item", str(incoming.get("entity_id"))) or {}
        if not item.get("pool"):
            raise Rejected("recount is only for a pool (FR-INV-35)")
    if entity_type == "found_report" and kind == "created":
        raise Rejected("found reports come from the public page")
    if kind == "photo_added":
        raise Rejected("photos are uploaded, not pushed")
    if entity_type == "code":
        if kind == "created":
            raise Rejected("codes come from printed sheets")
        if kind == "code_bound":
            code = derived.get_entity(conn, "code", str(incoming.get("entity_id")))
            if code is None:
                raise Rejected("not one of our codes")
            if code.get("item_id") is not None:
                raise Rejected("this code is already on an item")
            payload = incoming.get("payload")
            named = payload.get("item_id") if isinstance(payload, dict) else None
            target = derived.get_entity(conn, "item", str(named)) or {}
            if target.get("pool"):
                raise Rejected("a pool has no code (FR-INV-34)")
            if target.get("generic"):
                raise Rejected("a generic item takes no code; put it on a unit (FR-INV-21)")
        if kind == "code_released":
            code = derived.get_entity(conn, "code", str(incoming.get("entity_id")))
            if code is None:
                raise Rejected("not one of our codes")
            if code.get("item_id") is None:
                raise Rejected("this code is not on anything")


def _check_drift(conn: sqlite3.Connection, event: events.Event, measured_offset: int, now: int) -> None:
    """The offset the device recorded under should match what we measure now. If not, its clock moved."""
    drift = measured_offset - event.clock_offset
    if abs(drift) > DRIFT_THRESHOLD_MS:
        add_flag(
            conn,
            event.id,
            "clock_drift",
            {"recorded_offset": event.clock_offset, "measured_offset": measured_offset, "drift": drift},
            now,
        )


def pull(
    conn: sqlite3.Connection,
    principal: Principal,
    cursor: int,
    now: int | None = None,
    log: str | None = None,
) -> dict[str, Any]:
    """Events after a cursor, in seq order. The device calls again until it gets none.

    `log` is the log the device's snapshot came from, when it knows one.
    """
    _require_active(principal)
    now = now_ms() if now is None else now
    if cursor < 0:
        raise BadRequest("since must be a non-negative integer")
    if log is not None and log != log_id(conn):
        raise Rebootstrap("this is a different database; start again from a snapshot")

    conn.execute("BEGIN")
    try:
        last = conn.execute("SELECT coalesce(max(seq), 0) FROM events").fetchone()[0]
        if cursor > last:
            raise Rebootstrap("cursor is ahead of the log; the database was probably restored from backup")
        if cursor < last:
            if cursor == 0:
                base = conn.execute("SELECT min(received_at) FROM events").fetchone()[0]
            else:
                base = conn.execute(
                    "SELECT received_at FROM events WHERE seq <= ? ORDER BY seq DESC LIMIT 1", (cursor,)
                ).fetchone()[0]
            if base < now - RETENTION_MS:
                raise Rebootstrap("cursor is older than the retention window")
        page = events.since(conn, cursor, PAGE_SIZE)
    finally:
        conn.execute("COMMIT")

    return {
        "events": [asdict(e) for e in page],
        "cursor": page[-1].seq if page else cursor,
        "log_id": log_id(conn),
        "server_time": now,
    }
